### PR TITLE
Ms 5061 add ttl job results

### PR DESCRIPTION
**Pull request by dwelch-r7**
First stab at adding a ttl for the job results initiated from an RPC request

I was thinking it made a lot more sense for the calling service to define how it wanted to handle the jobs. This setup could allow for future implementations that for example utilise redis to store results or push results directly back out to the client using websockets bypassing the need to store them on the framework side at all.

### CODE REVIEW

File: lib/msf/core/rpc/v10/job_state_tracker.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+class JobStateTracker
+
```
Comment (AlanFoster) [inline lib/msf/core/rpc/v10/job_state_tracker.rb]: iirc we'll want to include the MonitorMixin here too

https://ruby-doc.org/stdlib-2.6.1/libdoc/monitor/rdoc/MonitorMixin.html#module-MonitorMixin-label-Simple+Class+include

File: spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        expect(@job_state_tracker.finished? @job_id).to eq(true)
+      end
+
+      context "we wait out the ttl" do
+        before(:each) do
+          sleep 0.1
+        end
+
+        it 'should not show as finished' do
+          expect(@job_state_tracker.finished? @job_id).to eq(false)
```
Comment (AlanFoster) [inline spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb]: iirc predicate methods exposed on objects can be called directly using rspec:

```ruby
expect(@job_state_tracker).to be_finished
```

https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/predicate-matchers

Which should hopefully improve readability 🎉

File: spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,159 @@
+require 'spec_helper'
+
+require 'msf/core/rpc/v10/job_state_tracker'
+
+RSpec.describe JobStateTracker do
+
+  context "With a 0.1 second ttl " do
```
Comment (AlanFoster) [inline spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb]: I'd be good to nuke this chunk of tests; ActiveSupport's MemoryStore is probably pretty well tested already, so there's not much need for us to duplicate that effort! 👀 

If you wanted to be sure that the `result_ttl` is passed correctly as a parameter to the memory store though, I believe you could do something like:

```ruby
before(:each) do
    allow(ActiveSupport::Cache::MemoryStore).to receive(:new).and_call_original

    JobStateTracker.new(result_ttl=0.1)
end

it { expect(ActiveSupport::Cache::MemoryStore).to have_received(:new).with(expires_in: 0.1) }
```

File: spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        end
+
+        it 'should not show as finished' do
+          expect(@job_state_tracker.finished? @job_id).to eq(false)
+        end
+      end
+    end
+  end
+
+  context "With default options" do
```
Comment (AlanFoster) [inline spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb]: What are your thoughts on using rspec's `let` syntax here:

```ruby
context "with default options" do
    let(:job_state_tracker) { described_class.new }
    let(:job_id) { "1234" }

    describe "#waiting" do
       before(:each) do
          job_state_tracker.waiting(job_id)
       end
      it { expect(job_state_tracker).to have_waiting job_id }
      it { expect(job_state_tracker).not_to have_running job_id }    
      it { expect(job_state_tracker).not_to have_finished job_id }    
    end
end
```

Extra notes:
- The `#waiting` notation comes from http://www.betterspecs.org/#describe
- The `described_class` comes from the argument given to `Rspec.describe` on line 5 - https://relishapp.com/rspec/rspec-core/docs/metadata/described-class
- I haven't tested any of this, your mileage may vary

File: lib/msf/core/rpc/v10/job_state_tracker.rb (hunk 1 above)
Comment (dwelch-r7) [reply lib/msf/core/rpc/v10/job_state_tracker.rb]: yup, 100% will do

File: spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb (hunk 2 above)
Comment (dwelch-r7) [reply spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb]: ah nice, didn't know about that, looks like a good change

File: spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb (hunk 3 above)
Comment (dwelch-r7) [reply spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb]: yea I really wasn't happy with this, but I'm not strictly speaking testing the MemoryStore so much as I'm trying to test the jobtrackers ability to TTL something, if that makes sense? 
I suppose making sure the memory store gets the right value passed in _should_ do for that though, cheers

File: spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb (hunk 4 above)
Comment (dwelch-r7) [reply spec/lib/msf/core/rpc/v10/job_state_tracker_spec.rb]: I'll poke about at it, it certainly looks nicer

File: lib/msf/base/simple/exploit.rb [hunk 5]
```diff
@@ -220,15 +220,12 @@ def self.job_check_proc(ctx)
...
+      mod.framework.job_state_tracker.start run_uuid
       result = mod.has_check? ? mod.check : Msf::Exploit::CheckCode::Unsupported
-      mod.framework.results[run_uuid] = {result: result}
+      mod.framework.job_state_tracker.completed(run_uuid, result)
     rescue => e
-      mod.framework.results[run_uuid] = {error: e.to_s}
+      mod.framework.job_state_tracker.failed(run_uuid, e)
       mod.handle_exception e
-    ensure
-      mod.framework.running.delete run_uuid
```
Comment (adfoster-r7) [inline lib/msf/base/simple/exploit.rb]: Do we need this still? 👀

Comment (dwelch-r7) [reply lib/msf/base/simple/exploit.rb]: it's handled in the job tracker now so don't think we should need it here

Comment (dwelch-r7) [reply lib/msf/base/simple/exploit.rb]: please double check my logic on that though, but I feel like it's the kind of thing that should be abstracted into the tracker

File: lib/msf/base/simple/job_state_tracker.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+  def start(id)
+    running << id
+    ready.delete(id)
+  end
+
+  def completed(id, result, ttl=nil)
+    begin
+      # ttl of nil means it will take the default expiry time
+      self.results_size += 1 if results.write(id, {result: result}, ttl)
```
Comment (AlanFoster) [inline lib/msf/base/simple/job_state_tracker.rb]: Correct me if i'm wrong, but will the value of `results_size` be wrong when the memory store expires its values behind the scenes?

Comment (dwelch-r7) [reply lib/msf/base/simple/job_state_tracker.rb]: you are 100% correct...I...don't have a way to solve that problem, not without reaching into places I shouldn't be in the cache

**Review by adfoster-r7 — APPROVED** [review]
Awesome! 🎉

**Comment by adfoster-r7** [comment]
# Release notes

Adds support for job results to be deleted after a period of 5 minutes of being un-acked by a json rpc client

**Pull request merged**